Switch to new Chrome headless mode

colinrotherham opened the pull request:
Following https://github.com/alphagov/govuk-frontend/pull/3542 just a quick PR for:

* https://github.com/alphagov/govuk-frontend/issues/3541

For more info on our headless mode deprecation issues see:

* https://github.com/puppeteer/puppeteer/issues/10070

colinrotherham commented:
Timing stats for `{ headless: true }` (before) versus `{ headless: 'new' }` (after)

## Before

**Accessibility tests:** 1m 57s
**Behaviour tests:** 41s
**Component tests:** 59s
**Percy screenshots:** 54s

## Before

**Accessibility tests:** 2m 27s
**Behaviour tests:** 46s
**Component tests:** 1m 4s
**Percy screenshots:** 1m 2s

36degrees commented:
Do you know if the Puppeteer team are aware that tests running under the new headless modes are slower? Is it something they're working to improve?

Sharding the accessibility tests is a neat idea (TIL that's a thing!) but I feel like our GitHub Actions are starting to get a little complex. Wondering if we should either wait until the performance of the new headless mode improves, or just accept that our tests will be a little bit slower for a while?

colinrotherham commented:
> Do you know if the Puppeteer team are aware that tests running under the new headless modes are slower? Is it something they're working to improve?

From [some of the comments](https://github.com/puppeteer/puppeteer/issues/10109#issuecomment-1532943513) they're aware of it (approx [30% slower](https://github.com/puppeteer/puppeteer/issues/10071#issuecomment-1521215003)), but appreciate it's now the same full browser for both headless and headful modes

> Sharding the accessibility tests is a neat idea (TIL that's a thing!) but I feel like our GitHub Actions are starting to get a little complex. Wondering if we should either wait until the performance of the new headless mode improves, or just accept that our tests will be a little bit slower for a while?

Ha that's great. I'd always wanted to try `--shard` out and this seemed perfect for it. That said, I'm more than happy to drop the commit (especially as it changes required status checks again!) and even hold off ["new headless"](https://developer.chrome.com/articles/new-headless/) for a while

For context this PR was thinking our [`{ headless: 1 }` loophole to silence the deprecation warning](https://github.com/alphagov/govuk-frontend/pull/3542) would be closed:

* https://github.com/puppeteer/puppeteer/issues/10070

But it's still holding strong

colinrotherham commented:
I've rebased with https://github.com/alphagov/govuk-frontend/pull/3491 and removed the `jest --shard` commit

No rush to approve this one but good to know it works when we need it

OrKoN commented:
The second timing block in https://github.com/alphagov/govuk-frontend/pull/3543#issuecomment-1521486713 is the "After" one (both say "before")?

colinrotherham commented:
> The second timing block in [#3543 (comment)](https://github.com/alphagov/govuk-frontend/pull/3543#issuecomment-1521486713) is the "After" one (both say "before")?

Ahh thanks, updated!

colinrotherham commented:
@36degrees I've gathered timing stats again after rebasing with Puppeteer v20:

* https://github.com/alphagov/govuk-frontend/pull/3589

The switch to [Chrome for Testing](https://goo.gle/chrome-for-testing) (not Chromium) has closed the gap 🙌

colinrotherham commented:
Although since the ~600 accessibility tests merged we've drifted a lot from the 2m 30s sweet spot we once had

romaricpascal reviewed (approved):
The timings now look pretty much the same as on main, let's merge this 🥳 ⛵
